Sources/NIO/ByteBuffer-aux.swift

```diff
@@ -431,4 +431,283 @@ extension ByteBuffer {
+    public init(string: String) {
+        self = ByteBufferAllocator().buffer(string: string)
+    }
+
+    /// Create a fresh `ByteBuffer` containing the bytes of the `string` encoded as UTF-8.
+    ///
+    /// This will allocate a new `ByteBuffer` with enough space fitting `string` and potentially some extra space using
+    /// the default allocator.
+    ///
+    /// - info: If you have access to a `Channel`, `ChannelHandlerContext`, or `ByteBufferAllocator` it is
+    ///         recommended using `channel.allocator.buffer(substring:)`. Or if you want to write multiple items into
+    ///         the buffer use `channel.allocator.buffer(capacity: ...)` to allocate a `ByteBuffer` of the right
+    ///         size followed by a `writeSubstring` instead of using this method. This allows SwiftNIO to do
+    ///         accounting and optimisations of resources acquired for operations on a given `Channel` in the future.
+    public init(substring string: Substring) {
+        self = ByteBufferAllocator().buffer(substring: string)
+    }
+
+    /// Create a fresh `ByteBuffer` containing the bytes of the `string` encoded as UTF-8.
+    ///
+    /// This will allocate a new `ByteBuffer` with enough space fitting `string` and potentially some extra space using
+    /// the default allocator.
+    ///
+    /// - info: If you have access to a `Channel`, `ChannelHandlerContext`, or `ByteBufferAllocator` it is
+    ///         recommended using `channel.allocator.buffer(staticString:)`. Or if you want to write multiple items into
+    ///         the buffer use `channel.allocator.buffer(capacity: ...)` to allocate a `ByteBuffer` of the right
+    ///         size followed by a `writeStaticString` instead of using this method. This allows SwiftNIO to do
+    ///         accounting and optimisations of resources acquired for operations on a given `Channel` in the future.
+    public init(staticString string: StaticString) {
+        self = ByteBufferAllocator().buffer(staticString: string)
+    }
+
+    /// Create a fresh `ByteBuffer` containing the `bytes`.
+    ///
+    /// This will allocate a new `ByteBuffer` with enough space fitting `bytes` and potentially some extra space using
+    /// the default allocator.
+    ///
+    /// - info: If you have access to a `Channel`, `ChannelHandlerContext`, or `ByteBufferAllocator` it is
+    ///         recommended using `channel.allocator.buffer(bytes:)`. Or if you want to write multiple items into the
+    ///         buffer use `channel.allocator.buffer(capacity: ...)` to allocate a `ByteBuffer` of the right
+    ///         size followed by a `writeBytes` instead of using this method. This allows SwiftNIO to do
+    ///         accounting and optimisations of resources acquired for operations on a given `Channel` in the future.
+    @inlinable
+    public init<Bytes: Collection>(bytes: Bytes) where Bytes.Element == UInt8 {
```

Comment:
This is more constrained than the base implementation in NIO, which uses `Sequence`. Any reason for that?